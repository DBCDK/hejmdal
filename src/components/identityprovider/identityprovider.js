import path from 'path';
import sendfile  from 'koa-sendfile';

export async function initialize(ctx, next) {
  if (!ctx.state) {
    ctx.state = {}
  }
  return next();
}

export async function authenticate(ctx, next) {
  ctx.state = {};
  if (true) {
    const filepath = path.join(__dirname, '../../../static/identityprovider.html');
    await sendfile(ctx, filepath);
    if (!ctx.status) {
      ctx.throw(404)
    }
  }
  return next();
}

export function callback(ctx, next) {
  ctx.state.user = {
    id: ctx.query.id,
    type: ctx.params.type,
    attributes: {
      unilogin: 'UNItestuser'
    }
  };
  ctx.body = ctx.state;
  next();
}
