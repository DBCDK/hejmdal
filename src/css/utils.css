.buffer-right-20 {
  margin-right: 20px;
}

.buffer-vertical-5 {
  margin-bottom: 5px;
}

.buffer-vertical-10 {
  margin-bottom: 10px;
}

.align-right{
  text-align: right;
}

.center-text {
  text-align: center;
}
