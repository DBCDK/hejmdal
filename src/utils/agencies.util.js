/**
 * @file
 * Methods for dealing with agencies
 */

import {libraryListFromName} from '../components/OpenAgency/openAgency.client';

/**
 * @type {Array}
 */
let agencyList = [];
let uiBranchList = [];

/**
 *
 * @return {Array}
 */
export async function cacheAgencies(name = '') {
  agencyList = await libraryListFromName(name);
}

/**
 *
 * @return {Array}
 */
export async function getListOfAgenciesForFrontend() {
  await setUiList();
  return uiBranchList;
}

/**
 * Return the name of the agency if found in agencyList
 *
 * @param agencyId
 * @returns {*}
 */
export async function getAgencyName(agencyId) {
  await setAgencyList();

  let name = '';
  if (agencyId) {
    name = 'Ukendt bibliotek: ' + agencyId;
    agencyList.forEach((agency) => {
      if (agency.branchId === agencyId) {
        name = agency.branchName;
      }
    });
  }
  return name;
}

export async function getAgency(agencyId) {
  await setAgencyList();

  let ret = {};
  if (agencyId) {
    agencyList.forEach((agency) => {
      if (agency.branchId === agencyId) {
        ret = agency;
      }
    });
  }
  return ret;

}

/**
 * agencyList setter
 */
async function setAgencyList() {
  if (!agencyList || !agencyList.length) {
    await cacheAgencies();
  }
}

/**
 * uiBranchList setter
 */
async function setUiList() {
  await setAgencyList();

  if (!uiBranchList || !uiBranchList.length) {
    uiBranchList = agencyList.map((branch) => {
      return {branchId: branch.branchId, name: branch.agencyName, hidden: branch.branchId};
    });
  }
}

/**
 * For testing purposes
 *
 * @param mockList
 */
export function mockSetAgencyList(mockList) {
  agencyList = mockList;
}
