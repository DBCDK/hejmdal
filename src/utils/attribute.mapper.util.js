/**
 * @file
 *
 *  Maps data from CULR (and IDprovider) as specified by the Smaug-setting for the given serviceClient
 *
 */

import {log} from './logging.util';
import {CONFIG} from './config.util';
import {mapFromCpr} from './cpr.util';
import {getInstitutionsForUser} from '../components/UniLogin/unilogin.component';
import {fetchDbcidpAuthorize, getDbcidpAgencyRightsAsFors, checkAgencyForProduct} from '../components/DBCIDP/dbcidp.client';

const removeAttributeAgencies = CONFIG.removeAttributeAgencies;

/**
 * Attribute mapper
 *
 * @param req
 * @param res
 * @param next
 * @returns {*}
 */
export default async function mapAttributesToTicket(req, res, next) {
  const state = req.getState();
  const user = req.getUser();

  if (state && state.serviceClient && state.culr) {
    const serviceAttributes = state.serviceClient.attributes;
    const culr = state.culr;
    const ticketAttributes = await mapCulrResponse(
      culr,
      serviceAttributes,
      user,
      state.serviceClient.id
    );
    req.setState({ticket: {attributes: ticketAttributes}});
  }

  await next();
}

/**
 * Maps the response from CULR according to the given array of fields
 *
 * @param {object} culr The CULR reponse object
 * @param {object} attributes The attributes object defined in Smaug
 * @param {object} user data returned by the idp
 * @param clientId
 * @param accessToken
 * @see ATTRIBUTES
 * @returns {Promise<{}>}
 */
export async function mapCulrResponse(
  culr,
  attributes,
  user,
  clientId,
  accessToken
) {
  let mapped = {};
  let cpr = user.cpr || null;
  let agencies = [];
  let fromCpr = {};
  let dbcidpAuthorize;
  let foundLoginAgency = false;

  log.debug('mapCulrResponse', culr);

  if (culr.accounts && Array.isArray(culr.accounts)) {
    culr.accounts.forEach(account => {
      if (account.userIdType === 'CPR' && !cpr) {
        cpr = account.userIdValue;
      }
      foundLoginAgency = foundLoginAgency || user.agency === account.provider;
      agencies.push({
        agencyId: account.provider,
        userId: account.userIdValue,
        userIdType: account.userIdType
      });
    });
  }
  if (!foundLoginAgency && user.agency && user.userId) {
    agencies.push({
      agencyId: user.agency,
      userId: user.userId,
      userIdType: user.cpr ? 'CPR' : 'LOCAL'
    })
  }

  if (cpr) {
    fromCpr = mapFromCpr(cpr);
  }

  const fields = Object.keys(attributes);

  await Promise.all(
    fields.map(async field => {  // eslint-disable-line complexity
      try {
        if (['dbcidp', 'dbcidpUniqueId', 'dbcidpUserInfo', 'dbcidpRoles'].includes(field)) {
          if (!dbcidpAuthorize) {
            dbcidpAuthorize = await fetchDbcidpAuthorize(accessToken, user);
          }
        }
        switch (field) {
          case 'birthDate':
          case 'birthYear':
          case 'gender':
            mapped[field] = fromCpr[field] || null;
            break;
          case 'cpr':
            mapped.cpr = cpr;
            break;
          case 'allAgencies':
          case 'allLibraries':
            mapped.agencies = agencies;
            break;
          case 'agencies':
          case 'libraries':
            mapped.agencies = [];
            agencies.forEach((agency) => {
              if (!removeAttributeAgencies.includes(agency.agencyId)) {
                mapped.agencies.push(agency);
              }
            });
            break;
          case 'blocked':
            mapped.blocked = culr.blocked;
            break;
          case 'userPrivilege':
            mapped.userPrivilege = [].concat(culr.userPrivilege ?? []);  // force array
            break;
          case 'municipality':
            mapped.municipality = culr.municipalityNumber || null;
            break;
          case 'municipalityAgencyId':
            mapped.municipalityAgencyId = culr.municipalityAgencyId || null;
            break;
          case 'uniloginId':
            mapped.uniloginId = user.uniloginId || null;
            break;
          case 'userId':
            mapped.userId = user.userId;
            break;
          case 'pincode':
            mapped.pincode = user.pincode;
            break;
          case 'wayfId':
            mapped.wayfId = user.wayfId ? user.wayfId : null;
            break;
          case 'uniqueId':
            mapped.uniqueId = culr.culrId;
            break;
          case 'uniLoginInstitutions': {
            mapped.uniLoginInstitutions =
              (user.uniloginId &&
                (await getInstitutionsForUser(user.uniloginId))) ||
              [];
            break;
          }
          case 'netpunktAgency':
            mapped.netpunktAgency = user.agency || null;
            break;
          case 'forsrights':
            mapped.forsrights = await getDbcidpAgencyRightsAsFors(accessToken, user);
            break;
          case 'dbcidp':
            mapped.dbcidp = [{agencyId: user.agency, rights: dbcidpAuthorize.rights || {}}];
            break;
          case 'dbcidpUniqueId':
            mapped.dbcidpUniqueId = dbcidpAuthorize.guid;
            break;
          case 'dbcidpUserInfo':
            mapped.dbcidpUserInfo = dbcidpAuthorize.userInfo;
            break;
          case 'dbcidpRoles':
            mapped.dbcidpRoles = dbcidpAuthorize.roles;
            break;
          case 'agencyRights':
            mapped.agencyRights = await getAgencyProduct(attributes[field], user.agency);
            break;
          case 'municipalityAgencyRights':
            mapped.municipalityAgencyRights = await getAgencyProduct(attributes[field], culr.municipalityAgencyId);
            break;
          case 'idpUsed':
            mapped.idpUsed = user.userType;
            break;
          default:
            log.warn('Cannot map attribute: ' + field);
            break;
        }
      } catch (error) {
        log.error('Failed to map attribute: ' + field, {
          error: error.message,
          stack: error.stack
        });
      }
    })
  );

  return mapped;
}

/**
 *
 * @param products
 * @param agency
 * @returns {Promise<*[]>}
 */
async function getAgencyProduct(products, agency) {
  const res = [];
  if (products.length && agency) {
    await Promise.all(products.map(async (product) => {
      const hasProduct = await checkAgencyForProduct(agency, product);
      res.push({[product]: hasProduct});
    }));

  }
  return res;
}
