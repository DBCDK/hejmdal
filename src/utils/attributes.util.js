/**
 * @file
 *
 * Describe the attributes supported by hejmdal.
 *
 * Each service specified in Smaug, can ask for one or more of these attributes.
 * The values of the attributes are fetched from CULR and DBCIDP and/or the chosen IDP
 *
 */

export const ATTRIBUTES = {
  cpr: {
    name: 'CPR-nummer',
    description: 'Brugerens CPR-nummer'
  },
  birthDate: {
    name: 'Fødselsdato',
    description: 'Fødselsdato - 4 første cifre af CPR-nummer'
  },
  birthYear: {
    name: 'Fødselsår',
    description: '4 cifret fødselsår - taget fra CPR-nummer'
  },
  gender: {
    name: 'Køn',
    description: 'Brugerens køn, m (for male) eller f (for female)'
  },
  allLibraries: {
    name: 'Biblioteker',
    description: 'En liste over de biblioteker som kender brugeren inklusiv interne biblioteker'
  },
  libraries: {
    name: 'Biblioteker',
    description: 'En liste over de biblioteker som kender brugeren'
  },
  municipality: {
    name: 'Kommunenummer',
    description: '3 cifret kommunenummer' // http://www.linking.dk/lokalportaler/kommuner.html
  },
  uniloginId: {
    name: 'UNI-login brugernavn',
    description: 'Brugerens identifikation hos UNI-Login'
  },
  userId: {
    name: 'Biblioteks bruger-id',
    description: 'Brugerens identitet på biblioteket - oftest CPR-nummer'
  },
  wayfId: {
    name: 'WAYF id',
    description: 'Brugerens identifikation hos WAYF'
  },
  uniqueId: {
    name: 'Bibliotekslogin id',
    description: 'Anonymiseret identifikation hos Bibliotekslogin'
  },
  authenticatedToken: {
    skipConsent: true
  },
  netpunktAgency: {},
  forsrights: {
    name: 'FORS-like Rights for agency',
    description: 'FORS-like produktrettigheder for det aktuelle agency'
  },
  dbcidp: {
    name: 'Rights for agency',
    description: 'Produktrettigheder for det aktuelle agency'
  },
  dbcidpUniqueId: {
    name: 'dbcidp unique id',
    description: 'unique id for user in dbcidp'
  },
  dbcidpUserInfo: {
    name: 'dbcidp user info',
    description: 'user info in dbcidp, name, email and phone'
  },
  dbcidpRoles: {
    name: 'dbcidp user roles',
    description: 'user roles in dbcidp'
  },
  blocked: {},
  userPrivilege: {},
  idpUsed: {
    name: 'Valgt IDP',
    description: 'Navn på den IDP der er logget ind med'
  },
  municipalityAgencyId: {},
  municipalityAgencyRights: {
    name: 'produkter i DBCIDP',
    description: 'Liste (array) af produkter som ønskes tjekket i DBCIDP for hjemhørs agency'
  },
  agencyRights: {
    name: 'produkter i DBCIDP',
    description: 'Liste (array) af produkter som ønskes tjekket i DBCIDP for det aktuelle agency'
  },
  tokenUser: {},
  uniLoginInstitutions: {
    name: 'List of unilogin institutions a user is connected to'
  }
};
