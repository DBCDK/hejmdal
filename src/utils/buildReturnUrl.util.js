/**
 * Build a returnUrl from state.
 *
 * @param state
 * @param querystring
 * @returns {string}
 */
export default function buildReturnUrl(state, queryObject = {}) {
  let host = state.serviceClient.urls.host;
  let path = state.returnUrl || state.serviceClient.urls.returnUrl;
  let querystring = objectToQueryString(queryObject);

  if (path.charAt(0) !== '/') {
    path = '/' + path;
  }

  if (host.charAt(host.length - 1) === '/') {
    host = host.substr(0, host.length - 1);
  }

  return `${host}${path}${querystring}`;
}

/**
 * convert object to url query string.
 * @param queryObject
 * @returns {string}
 */
function objectToQueryString(queryObject) {
  let queryString = '';
  const keys = Object.keys(queryObject);

  if (keys.length) {
    queryString = '?' + keys.map(key => {
      const value = encodeURIComponent(queryObject[key]);
      return `${key}=${value}`;
    }).join('&');
  }

  return queryString;
}
