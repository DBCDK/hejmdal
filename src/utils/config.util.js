/**
 * @file
 * Config mapper that maps environment variables to the exportet CONFIG object.
 * A validateConfig method that validates the values found in the CONFIG object and throws an Error upon invalid values.
 */


export const CONFIG = {
  app: {
    env: process.env.NODE_ENV,
    port: Number(process.env.PORT)
  },
  postgres: {
    client: 'postgresql',
    connection: {
      host: process.env.HEJMDAL_DB_HOST,
      database: process.env.HEJMDAL_DB_NAME,
      user: process.env.HEJMDAL_DB_USER,
      password: process.env.HEJMDAL_DB_USER_PASSWORD,
      charset: 'utf8'
    },
    pool: {
      min: Number(process.env.HEJMDAL_DB_CONNECTIONS_POOL_MIN),
      max: Number(process.env.HEJMDAL_DB_CONNECTIONS_POOL_MAX)
    },
    migrations: {
      tableName: 'knex_migrations'
    }
  },
  log: {
    level: process.env.LOG_LEVEL,
    pretty: process.env.PRETTY_LOG === '1'
  },
  session: {
    life_time: Number(process.env.SESSION_LIFE_TIME)
  },
  hash: {
    shared: process.env.HASH_SHARED
  },
  garbageCollect: {
    ticket: {
      divisor: Number(process.env.GC_TICKET_DIVISOR) || 1000,
      seconds: Number(process.env.GC_TICKET_SECONDS) || 3600   // 60 * 60
    },
    session: {
      divisor: Number(process.env.GC_SESSION_DIVISOR) || 1000,
      seconds: Number(process.env.GC_SESSION_SECONDS) || 2678400  // 31 * 24 * 60 * 60
    }
  },
  borchk: {
    uri: process.env.BORCHK_URI
  },
  smaug: {
    uri: process.env.SMAUG_URI
  },
  mock_externals: {
    ticket: process.env.MOCK_TICKET_STORAGE || false,
    consent: process.env.MOCK_CONSENT_STORAGE || false,
    smaug: process.env.MOCK_SMAUG !== '0',
    borchk: process.env.MOCK_BORCHK !== '0'
  }
};

/**
 * Recursive functon that validates that all params in the above CONFIG object is set.
 * Number are validated to be non-NaN numbers.
 *
 * @param {Object} config
 */
export function validateConfig(config = CONFIG) {
  for (let key in config) {
    if (typeof config[key] === 'object') {
      validateConfig(config[key]);
    }
    else {
      if (config[key] === undefined) { // eslint-disable-line no-undefined
        throw Error(`${key} was not specified in config. See https://github.com/DBCDK/hejmdal#environment-variabler`);
      }
      if (typeof config[key] === 'number' && Number.isNaN(config[key])) {
        throw Error(`${key}: expected NaN to be a number. See https://github.com/DBCDK/hejmdal#environment-variabler`);
      }
    }
  }
}
