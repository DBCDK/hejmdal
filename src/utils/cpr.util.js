/**
 * @file
 */

/**
 * Creates birthdate, birthYear and gender from CPR
 * @param cpr
 * @returns {{}}
 */
export function mapFromCpr(cpr) {
  const ret = {};
  if (isValidCpr(cpr)) {
    ret.birthDate = cpr.substr(0, 4);
    ret.birthYear = addMilenium(cpr.substr(4, 2), cpr.substr(6, 1));
    ret.gender = cpr.substr(9, 1) % 2 ? 'm' : 'f';
  }
  return ret;
}

/**
 *  Add milenium to 2 digit year as specified by https://da.wikipedia.org/wiki/CPR-nummer
 *
 * @param year - 2 digit year
 * @param seven - digit number 7 in the cpr
 */
export function addMilenium(year, seven) {
  switch (seven) {
    case '0':
    case '1':
    case '2':
    case '3':
      return '19' + year;
    case '4':
    case '9':
      return (year >= 37 ? '19' : '20') + year;
    case '5':
    case '6':
    case '7':
    case '8':
      return (year >= 58 ? '18' : '20') + year;
    default:
      return '??' + year;
  }
}
/**
 * Check for validity of a cpr
 *
 * @param cpr
 * @returns {boolean|Date|boolean}
 */
export function isValidCpr(cpr) {
  return isNumeric(cpr) && isValidDate(cpr) && cpr.length === 10;
}

/**
 *
 * @param n
 * @returns {boolean}
 */
export function isNumeric(n) {
  return !isNaN(parseFloat(n)) && isFinite(n);
}

/**
 * Check if first 6 digits contains a valid date
 * @param ddmmyy
 * @returns {Date|boolean}
 */
export function isValidDate(ddmmyy) {
  const mm = ddmmyy.substr(2, 2) || '00';
  const d = new Date(ddmmyy.substr(4, 2), mm - 1, ddmmyy.substr(0, 2));
  return d && d.getMonth() + 1 === parseInt(mm, 10);
}
