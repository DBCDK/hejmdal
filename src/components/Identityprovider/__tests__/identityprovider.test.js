import {assert} from 'chai';
import sinon from 'sinon';
import {authenticate, identityProviderCallback} from '../identityprovider.component';
import {createHash} from '../../../utils/hash.utils';
import {mockContext} from '../../../utils/test.util';
import moment from 'moment';
import {md5} from '../../../utils/hash.utils';
import {CONFIG} from '../../../utils/config.util';

describe('test authenticate method', () => {
  const next = () => {};
  let ctx;
  let sandbox;

  beforeEach(() => {
    ctx = mockContext();
    sandbox = sinon.sandbox.create();
  });

  afterEach(() => {
    sandbox.restore();
  });

  it('Should return content page', async() => {
    ctx.setState({
      serviceClient: {
        identityProviders: ['borchk', 'unilogin'],
        urls: {
          host: 'https://service_client_url'
        }
      }
    });
    ctx.render = sandbox.mock();
    await authenticate(ctx, ctx, next);
    assert.equal(ctx.status, 200);
    sandbox.restore();
  });

  it('Should render error page', async() => {
    const spy = sandbox.spy(ctx, 'render');
    ctx.setState({serviceClient: {identityProviders: ['invalid provider']}});
    assert.isFalse(spy.called);

    await authenticate(ctx, ctx, next);

    assert.isTrue(spy.called);
    assert.isString(spy.args[0][1].error);
    assert.isObject(spy.args[0][1].link);
  });
});

describe('test identityProviderCallback method', () => {
  let ctx;

  beforeEach(() => {
    ctx = mockContext('qwerty', 'some_url', {
      params: {},
      query: {
        id: 'testId'
      }
    });
    ctx.params.state = createHash(ctx.getState().stateHash);
  });

  const next = () => {
  };

  it('Should add unilogin user to context', async() => {
    ctx.params.type = 'unilogin';
    const user = 'test1234';
    const timestamp = moment().utc().format('YYYYMMDDHHmmss');
    const auth = md5(timestamp + CONFIG.unilogin.secret + user);
    ctx.query = {
      auth: auth,
      timestamp: timestamp,
      user: user
    };

    const expected = {
      userId: 'test1234',
      userType: 'unilogin',
      identityProviders: ['unilogin']
    };

    await identityProviderCallback(ctx, ctx, next);

    assert.deepEqual(ctx.getUser(), expected);
  });

  it('Should add nemlogin user and cpr to context', async() => {
    ctx.params.type = 'nemlogin';
    const expected = {
      userId: '0102030405',
      cpr: '0102030405',
      userType: 'nemlogin',
      identityProviders: ['nemlogin']
    };
    await identityProviderCallback(ctx, ctx, next);
    assert.deepEqual(ctx.getUser(), expected);
  });

  it('Should add borchk user to context', async() => {
    ctx.params.type = 'borchk';
    ctx.fakeBorchkPost = {userId: 'testId', pincode: 'testPincode', libraryId: '710100'};
    const expected = {
      userId: 'testId',
      cpr: null,
      userType: 'borchk',
      libraryId: '710100',
      pincode: 'testPincode',
      identityProviders: ['borchk'],
      userValidated: true
    };
    await identityProviderCallback(ctx, ctx, next);
    assert.deepEqual(ctx.getUser(), expected);
  });
});
