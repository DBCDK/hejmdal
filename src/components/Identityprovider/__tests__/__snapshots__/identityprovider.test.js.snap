// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test authenticate method Should render error page 1`] = `
[MockFunction] {
  "calls": [
    [
      "Error",
      {
        "error": "Der opstod en fejl som skyldes forkert konfiguration af Bibliotekslogin. Kontakt gerne en administrator på dit bibliotek og gør opmærksom på fejlen.",
        "link": {
          "href": "/login",
          "value": "Prøv igen",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
